internal/datanode/writebuffer/insert_buffer.go

```diff
@@ -87,40 +97,96 @@ func NewInsertBuffer(sch *schemapb.CollectionSchema) (*InsertBuffer, error) {
 	if estSize == 0 {
 		return nil, errors.New("Invalid schema")
 	}
-	buffer, err := storage.NewInsertData(sch)
-	if err != nil {
-		return nil, err
-	}
+
 	sizeLimit := paramtable.Get().DataNodeCfg.FlushInsertBufferSize.GetAsInt64()
+	chunkSize := paramtable.Get().DataNodeCfg.InsertBufferChunkSize.GetAsInt64()
+	// use size Limit when chunkSize not valid
+	if chunkSize <= 0 || chunkSize > sizeLimit {
+		log.Warn("invalidate chunk size, use insert buffer size", zap.Int64("chunkSize", chunkSize), zap.Int64("insertBufferSize", sizeLimit))
+		chunkSize = sizeLimit
+	}
 
-	return &InsertBuffer{
+	estRow := chunkSize / int64(estSize)
+	ib := &InsertBuffer{
 		BufferBase: BufferBase{
 			rowLimit:      noLimit,
 			sizeLimit:     sizeLimit,
 			TimestampFrom: math.MaxUint64,
 			TimestampTo:   0,
 		},
 		collSchema: sch,
-		buffer:     buffer,
-	}, nil
+		estRow:     estRow,
+		chunkSize:  chunkSize,
+		buffers:    make([]*InsertBufferChunk, 0, sizeLimit/chunkSize),
+	}
+	err = ib.nextBatch()
+	if err != nil {
+		return nil, err
+	}
+
+	return ib, nil
+}
+
+func (ib *InsertBuffer) nextBatch() error {
+	buffer, err := storage.NewInsertDataWithCap(ib.collSchema, int(ib.estRow))
+	if err != nil {
+		return err
+	}
+	ib.buffers = append(ib.buffers, &InsertBufferChunk{
+		BufferBase: BufferBase{
+			rowLimit:      ib.estRow,
+			sizeLimit:     ib.chunkSize,
+			TimestampFrom: math.MaxUint64,
+			TimestampTo:   0,
+		},
+		buffer: buffer,
+	})
+	return nil
+}
+
+func (ib *InsertBuffer) currentBuffer() *InsertBufferChunk {
+	idx := len(ib.buffers) - 1
+	if idx < 0 || ib.buffers[idx].IsFull() {
+		ib.nextBatch()
+		idx++
+	}
+	return ib.buffers[idx]
+}
+
+func (ib *InsertBuffer) buffer(inData *storage.InsertData, tr TimeRange, startPos, endPos *msgpb.MsgPosition) {
+	buffer := ib.currentBuffer()
+	storage.MergeInsertData(buffer.buffer, inData)
+	buffer.UpdateStatistics(int64(inData.GetRowNum()), int64(inData.GetMemorySize()), tr, startPos, endPos)
 }
 
 func (ib *InsertBuffer) Yield() *storage.InsertData {
 	if ib.IsEmpty() {
 		return nil
 	}
 
-	return ib.buffer
+	// avoid copy when there is only one buffer
+	if len(ib.buffers) == 1 {
+		return ib.currentBuffer().buffer
+	}
+	// no error assumed, buffer created before
+	result, _ := storage.NewInsertDataWithCap(ib.collSchema, int(ib.rows))
```

Comment:
maybe we don't really need to return a real *storage.InsertData ?
can we instead return a insertDataInterator? so we can avoid one extra copy?

Reply:
Good point. Working on it